### DISCUSSION TITLE
Access api dependencies in middleware

### DISCUSSION THREAD

**Discussion by pankhuriagarwal94** [fastapi/fastapi#8939] [Questions] [answered]
I want to access Api level dependency of cache in custom request class. Api is defined using router and custom request middleware class is defined APIRoute.

APIRoute class

    from typing import Callable

    from fastapi import Request, Response, Depends
    from fastapi.routing import APIRoute

    class RecordRequestResponseRoute(APIRoute):
        def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request)\
                -> Response:
            
            # need to access cache here 
            response: Response = await original_route_handler(request)
            return response

        return custom_route_handler

Router Api

    from fastapi import Response, Depends, APIRouter, BackgroundTasks
    import fastapi_plugins
    import aioredis

    router = APIRouter(route_class=RecordRequestResponseRoute)

    @router.get("/users", tags=["users"])
    async def match_frame(background_tasks: BackgroundTasks,
                      cache: aioredis.Redis = Depends(fastapi_plugins.depends_redis)):
         return {"success": True, "data": []}

I need to access cache in RecordRequestResponseRoute class. I tried using sub dependencies but that did not help. What will be the correct way to do it?

**Comment by ginomempin** [comment, marked as answer]
I'm having the same issue. The only way I found so far is by creating the dependency (ex. `cache`) for yourself inside the `custom_route_handler`, using parts of the `request`. But, fair warning, this is **NOT** a good thing to do (I don't consider this a solution), as this duplicates the `Depends` functionality and I did not end up using it.

I don't have Redis cache so here's a simplified example:

```
class Data:
    def __init__(self, x: int = 0):
        self.x = x

class CustomRoute(APIRoute):
    def get_route_handler(self) -> Callable:
        original_route_handler = super().get_route_handler()

        async def custom_route_handler(request: Request) -> Response:
            for dep in self.dependant.dependencies:
                if dep.name != 'data':      # same name as in the route
                    continue
                dep_cls = dep.call           # <class 'app.Data'>
                data = dep_cls(request.query_params.get('x', 0))
                # print(data.x) or logging.debug(data.x)
            response: Response = await original_route_handler(request)
            return response

        return custom_route_handler

custom_router = APIRouter(route_class=CustomRoute)

@custom_router.get('/')
def info(x: int = 0, data: Data = Depends(Data)):
    print(data.x)
    return {'x': data.x}
```

Here, you can access the routes **dependency class** and the **query params**,.... then create the dependency object yourself. You should get the same print result as accessing the endpoint at `http://localhost:8000?x=33`. 

This duplicates the work that the `original_route_handler` will also do when you call `await original_route_handler` to get the response. This also won't work if the dependency is created using other objects that are not part of the route params or from other sub-dependencies.

Anyway, hoping for a better solution or approach than this.
